File: javafx-application/src/main/kotlin/dev/kkorolyov/pancake/javafx/render/JavaFXImage.java
package dev.kkorolyov.pancake.javafx.render;

import dev.kkorolyov.pancake.platform.math.Vector;
import dev.kkorolyov.pancake.platform.media.graphic.Image;
import dev.kkorolyov.pancake.platform.media.graphic.RenderTransform;

/**
 * {@link Image} implemented through JavaFX.
 */
public class JavaFXImage extends Image {
	private final javafx.scene.image.Image image;
	private final EnhancedGraphicsContext g;

	/**
	 * Constructs a new JavaFX-backed image instance.
	 * @param image backing JavaFX image instance
	 * @param g backing graphics context
	 */
	public JavaFXImage(javafx.scene.image.Image image, EnhancedGraphicsContext g) {
		this.image = image;
		this.g = g;
	}

	@Override
	public void render(RenderTransform transform) {
		Vector origin = getViewport().getOrigin(image.getWidth(), image.getHeight());
		Vector size = getViewport().getSize(image.getWidth(), image.getHeight());

		// Rotate
		g
				.rotate(transform)
				.get()
				.drawImage(
						image,
						// Select source using viewport
						origin.getX(),
						origin.getY(),
						size.getX(),
						size.getY(),
						// Render using transform
						transform.getPosition().getX(),
						transform.getPosition().getY(),
						size.getX() * transform.getScale().getX(),
						size.getY() * transform.getScale().getY()
				);
	}
}

File: javafx-application/src/main/kotlin/dev/kkorolyov/pancake/javafx/render/JavaFxImage.kt
package dev.kkorolyov.pancake.javafx.render

import dev.kkorolyov.pancake.platform.math.Vector2
import dev.kkorolyov.pancake.platform.media.graphic.Image
import dev.kkorolyov.pancake.platform.media.graphic.RenderTransform
import javafx.scene.image.Image as FxImage

/**
 * [Image] implemented through JavaFX.
 */
class JavaFxImage(private val image: FxImage, private val g: EnhancedGraphicsContext) : Image() {
	override fun render(transform: RenderTransform) {
		val origin: Vector2 = viewport.getOrigin(image.width, image.height)
		val size: Vector2 = viewport.getSize(image.width, image.height)

		// Rotate
		g.run {
			rotate(transform, pad = true)

			get().drawImage(
				image,
				// Select source using viewport
				// Select source using viewport
				origin.x,
				origin.y,
				size.x,
				size.y,
				// Render using transform
				// Render using transform
				transform.position.x - (size.x / 2),
				transform.position.y - (size.y / 2),
				size.x * transform.scale.x,
				size.y * transform.scale.y
			)
		}
	}
}
